api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -25,55 +35,223 @@
+      if (currentFraction > threshold) {
+        // only alert if we have not done so previously
+        if (previousFraction <= threshold) {
+          notificationService.alertUserFreeTierDollarThreshold(
+              user, threshold, currentCost, remainingBalance);
+        }
+
+        // break out here to ensure we don't alert for lower thresholds
+        break;
+      }
+    }
+  }
+
+  /**
+   * Has this user passed a time threshold between this check and the previous run?
+   *
+   * <p>Compare this user's free credits timespan with that of the previous run, and trigger an
+   * alert if this is the run which pushed it over a threshold.
+   *
+   * @param user The user to check
+   * @param remainingDollarBalance The remaining dollar balance to this user, for reporting purposes
+   * @param currentCheckTime a fixed time common to all checks of this run, for comparison purposes
+   */
+  private void maybeAlertOnTimeThresholds(
+      DbUser user, double remainingDollarBalance, Instant currentCheckTime) {
+    final Instant userFreeCreditStartTime = user.getFirstRegistrationCompletionTime().toInstant();
+
+    final Instant previousCheckTime =
+        Optional.ofNullable(user.getLastFreeTierCreditsTimeCheck())
+            .map(Timestamp::toInstant)
+            .orElse(userFreeCreditStartTime);
+
+    // save current check time
+    user.setLastFreeTierCreditsTimeCheck(Timestamp.from(currentCheckTime));
+    userDao.save(user);
+
+    final Duration userFreeCreditDays = Duration.ofDays(getUserFreeTierDaysLimit(user));
+    final Duration currentTimeElapsed = Duration.between(userFreeCreditStartTime, currentCheckTime);
+    final Duration previousTimeElapsed =
+        Duration.between(userFreeCreditStartTime, previousCheckTime);
+
+    // can't use toDays() here because it truncates and we need sub-day resolution
+    final double currentFraction =
+        (double) currentTimeElapsed.toMillis() / userFreeCreditDays.toMillis();
+    final double previousFraction =
+        (double) previousTimeElapsed.toMillis() / userFreeCreditDays.toMillis();
+
+    final Instant userFreeCreditExpirationTime = userFreeCreditStartTime.plus(userFreeCreditDays);
+    final Duration timeRemaining = Duration.between(currentCheckTime, userFreeCreditExpirationTime);
+
+    for (final double threshold : alertThresholdsInDescOrder) {
+      if (currentFraction > threshold) {
+        // only alert if we have not done so previously
+        if (previousFraction <= threshold) {
+          notificationService.alertUserFreeTierTimeThreshold(
+              user,
+              timeRemaining.toDays(),
+              userFreeCreditExpirationTime.atZone(ZoneId.systemDefault()).toLocalDate(),
+              remainingDollarBalance);
+        }
+
+        // break out here to ensure we don't alert for lower thresholds
+        break;
+      }
+    }
+  }
+
+  // we set Workspaces to EXPIRED when their creators have exceeded their free credits
```

Comment:
It doesn't seem like we're doing this. Could you update the comment to `INACTIVE`?

Reply:
good catch, thanks